crates/uv/src/commands/project/version.rs

```diff
@@ -0,0 +1,593 @@
+    DiscoveryOptions, WorkspaceCache,
+    pyproject_mut::{DependencyTarget, PyProjectTomlMut},
+};
+use uv_workspace::{VirtualProject, Workspace};
+
+use crate::commands::pip::loggers::{DefaultInstallLogger, DefaultResolveLogger};
+use crate::commands::pip::operations::Modifications;
+use crate::commands::project::add::{AddTarget, PythonTarget};
+use crate::commands::project::install_target::InstallTarget;
+use crate::commands::project::lock::LockMode;
+use crate::commands::project::{
+    ProjectEnvironment, ProjectError, ProjectInterpreter, UniversalState, default_dependency_groups,
+};
+use crate::commands::{ExitStatus, diagnostics, project};
+use crate::printer::Printer;
+use crate::settings::{NetworkSettings, ResolverInstallerSettings};
+
+/// Display version information for uv itself (`uv self version`)
+pub(crate) fn self_version(
+    short: bool,
+    output_format: VersionFormat,
+    printer: Printer,
+) -> Result<ExitStatus> {
+    let version_info = uv_cli::version::uv_self_version();
+    print_version(version_info, None, short, output_format, printer)?;
+
+    Ok(ExitStatus::Success)
+}
+
+/// Read or update project version (`uv version`)
+#[allow(clippy::fn_params_excessive_bools)]
+pub(crate) async fn project_version(
+    value: Option<String>,
+    bump: Option<VersionBump>,
+    short: bool,
+    output_format: VersionFormat,
+    strict: bool,
+    project_dir: &Path,
+    package: Option<PackageName>,
+    dry_run: bool,
+    locked: bool,
+    frozen: bool,
+    active: Option<bool>,
+    no_sync: bool,
+    python: Option<String>,
+    install_mirrors: PythonInstallMirrors,
+    settings: ResolverInstallerSettings,
+    network_settings: NetworkSettings,
+    python_preference: PythonPreference,
+    python_downloads: PythonDownloads,
+    installer_metadata: bool,
+    concurrency: Concurrency,
+    no_config: bool,
+    cache: &Cache,
+    printer: Printer,
+    preview: PreviewMode,
+) -> Result<ExitStatus> {
+    // Read the metadata
+    let project = match find_target(project_dir, package.as_ref()).await {
+        Ok(target) => target,
+        Err(err) => {
+            // If strict, hard bail on failing to find the pyproject.toml
+            if strict {
+                return Err(err)?;
+            }
+            // Otherwise, warn and provide fallback to the old `uv version` from before 0.7.0
+            warn_user!(
+                "Failed to read project metadata ({err}). Running `{}` for compatibility. This fallback will be removed in the future; pass `--preview` to force an error.",
+                "uv self version".green()
+            );
+            return self_version(short, output_format, printer);
+        }
+    };
+
+    let mut toml = PyProjectTomlMut::from_toml(
+        project.pyproject_toml().raw.as_ref(),
+        DependencyTarget::PyProjectToml,
+    )?;
+
+    let pyproject_path = project.root().join("pyproject.toml");
+    let name = project.project_name().cloned();
```

Comment:
I think if we don't have a `name`, then we don't have a `[project]` table, so we don't have a version?